src/replay/consensus.zig

```diff
@@ -0,0 +1,1321 @@
+    );
+}
+
+/// Pushes a new vote transaction to the network and updates tower state.
+///
+/// - Generates a new vote transaction.
+/// - Updates the tower's last vote blockhash.
+/// - Creates a saved tower state.
+/// - Sends the vote operation to the voting sender.
+///
+/// Analogous to [push_vote](https://github.com/anza-xyz/agave/blob/ccdcdbe9b6ff7dbd583d2101fe57b7cc41a6f863/core/src/replay_stage.rs#L2775)
+fn pushVote(
+    replay_tower: *ReplayTower,
+) !void {
+
+    // TODO Transaction generation to be implemented.
+    // Currently hardcoding to non voting transaction.
+    const vote_tx_result: GenerateVoteTxResult = .non_voting;
+
+    switch (vote_tx_result) {
+        .tx => |vote_tx| {
+            _ = &vote_tx;
+            // TODO to be implemented
+        },
+        .non_voting => {
+            replay_tower.markLastVoteTxBlockhashNonVoting();
+        },
+        else => {
+            // Do nothing
+        },
+    }
+}
+
+/// Processes a new root slot by updating various system components to reflect the new root.
+///
+/// - Validates the new root exists and has a hash.
+/// - Gets all slots rooted at the new root.
+/// - Updates ledger state with the new roots.
+/// - Cleans up progress map for non-existent slots.
+/// - Updates fork choice with the new root.
+///
+/// Analogous to [check_and_handle_new_root](https://github.com/anza-xyz/agave/blob/ccdcdbe9b6ff7dbd583d2101fe57b7cc41a6f863/core/src/replay_stage.rs#L4002)
+fn checkAndHandleNewRoot(
+    allocator: std.mem.Allocator,
+    blockstore_reader: *BlockstoreReader,
+    slot_tracker: *SlotTracker,
+    progress: *ProgressMap,
+    fork_choice: *ForkChoice,
+    new_root: Slot,
+) !void {
+    // get the root bank before squash.
+    if (slot_tracker.slots.count() == 0) return error.EmptySlotTracker;
+    var root_tracker = slot_tracker.slots.get(new_root) orelse return error.MissingSlot;
+    const maybe_root_hash, var hash_lg = root_tracker.state.hash.readWithLock();
+    defer hash_lg.unlock();
+    const root_hash = maybe_root_hash.* orelse return error.MissingHash;
+
+    const rooted_slots = try slot_tracker.parents(allocator, new_root);
+    defer allocator.free(rooted_slots);
+
+    // TODO implement leader_schedule_cache.set_root.
+    // TODO have this a seperate function?
+    {
+        var lowest_cleanup_slot = RwMux(Slot).init(0);
+        var max_root = std.atomic.Value(Slot).init(0);
+        var registry = sig.prometheus.Registry(.{}).init(allocator);
+        defer registry.deinit();
+
+        var writer = LedgerResultWriter{
+            .allocator = allocator,
+            .db = blockstore_reader.db,
+            .logger = .noop,
+            .lowest_cleanup_slot = &lowest_cleanup_slot,
+            .max_root = &max_root,
+            .scan_and_fix_roots_metrics = try registry.initStruct(
+                sig.ledger.result_writer.ScanAndFixRootsMetrics,
+            ),
+        };
+
+        try writer.setRoots(rooted_slots);
```

Comment:
setRoots updates the state behind the `max_root` pointer in LedgerResultWriter. This is supposed to be the max_root state that's shared with the rest of the validator. That value is initialized in cmd.zig. But instead you're just initializing a value here and discarding it immediately after mutating it. So this code is not having the intended effect.

The safest bet is probably to add a dependency on LedgerResultWriter, initialize that struct using the same max_root as BlockstoreReader in the validator function of cmd.zig, and pass it down to here.

Eventually, I'd like to improve the ledger structs. The shared state is confusing, and the function's dependencies are overkill. But I think for now we can live with a centrally constructed LedgerResultWriter.

Reply:
[bc35b76](https://github.com/Syndica/sig/pull/762/commits/bc35b7695d2b345737a717e5f8d99cc3afdbd410)